### DISCUSSION TITLE
1bpp bmp image comes out as having 3 components per pixel

### DISCUSSION THREAD

**Discussion by ghost** [nothings/stb#1326] [Q&A] [answered]
I am reading a B&W bmp file with 1 bpp and I noticed that the `comp` argument returns after the call carrying the value `3`. It seems that the value is set at line `5509` in `stb/stb_image.h`:

````
   if (info.bpp == 24 && ma == 0xff000000)
      s->img_n = 3;
   else
      s->img_n = ma ? 4 : 3;
````

The else branch is taken and `ma` is `0`. Am I wrong to expect `comp` to be `1`? Thanks.

**Comment by ghost** [comment, marked as answer]
Alright, it looks like user error. The image has a bit-depth of 1 but ImageMagick reports an sRGB colorspace with three channels, RGB, each of depth 1. Bad input.

**Comment by rygorous** [comment]
I believe the way this works out is that 1bpp BMP files are taken to be paletted and those colors come from the palette.

**Reply by ghost** [reply]
That is correct. There is a palette with two colors, "white" and "black".
